grin/test/Test/EndToEnd.hs

```diff
@@ -0,0 +1,246 @@
+{-# LANGUAGE TypeFamilies, LambdaCase, TypeApplications, DeriveGeneric, ScopedTypeVariables #-}
+module Test.EndToEnd where
+
+import CLI.Lib (mainWithArgs)
+import Control.Arrow ((&&&))
+import Control.Exception (catch)
+import Control.Monad (forM, when)
+import Data.ByteString.Char8 (ByteString)
+import Data.Char (isDigit)
+import Data.IORef
+import Data.List (isSuffixOf)
+import Data.Maybe (fromMaybe)
+import Data.String (fromString)
+import Data.String.Utils (replace)
+import Data.Yaml as Yaml
+import GHC.Generics
+import GHC.IO.Handle
+import System.Directory
+import System.Directory (doesFileExist, removeDirectoryRecursive)
+import System.Directory (doesFileExist, removeFile)
+import System.Directory.Tree
+import System.FilePath.Posix
+import System.Posix.Redirect
+import System.Process
+import Test.Hspec.Core.Spec hiding (pending)
+import qualified Data.Map as Map
+
+
+data InputFile
+  = Binary  FilePath
+  | Textual FilePath
+  deriving (Eq, Show)
+
+inputToFilePath :: InputFile -> FilePath
+inputToFilePath = \case
+  Binary fp -> fp
+  Textual fp -> fp
+
+data CompilerTest
+  = PipelineTest
+    { compilerInput     :: InputFile  -- name.grin name.binary
+    , compilerOptions   :: [String]   -- input.opts
+    , compilerExpected  :: FilePath   -- input.expected
+    , compilerOutputExt :: FilePath   -- input.out -- defined in configuration
+    }
+  | EndToEndTest
+    { compilerInput :: InputFile
+    }
+
+-- | evaluatePipelineTest reads the input file, runs grin compiler using the
+-- defined options. All the $$$OUT$$$ template value replaced the output file.
+-- Only one output file should be configured in the options as we can define
+-- only one expected output. The output file should be text file of some sort.
+evaluatePipelineTest input options expected ext params actionWith progressCallback = do
```

Comment:
Why is the type signature missing? Is it too complicated?

Reply:
It is a local helper function called only once. If you are really think a type signature is necessary I'll add them, but at the time of the writing I didn't feel the need of adding the type signatures.

Reply:
In my opinion, it is always good to have type signatures for top-level functions. Makes the code more readable, makes the docs more understandable.

Reply:
Than I'll add that ugly ones :D